### ISSUE TITLE
"docker build -f FILE_NAME" requires full path to the file

### ISSUE THREAD

**Issue by wzheng2310** [docker/docker#14339]
Hi, 

When I run this from command line:
docker build -t XXX --file Dockerfile /full/path/to/directory/containing/the/given/dockerfile

I got this error:
The Dockerfile (Dockerfile) must be within the build context (/full/path/to/directory/containing/the/given/dockerfile/)

I have to give full path to the filename, i.e.:
docker build -t XXX --file /full/path/to/directory/containing/the/given/dockerfile/Dockerfile /full/path/to/directory/containing/the/given/dockerfile

This does not make much sense to me. The path to the directory (i.e. the build context) is already given, the file name should be assumed to be under that directory.

**Comment by wzheng2310** [comment]
This is a duplicate of issue 11289.

**Comment by rrnewton** [comment]
This is pretty unclear to me.  Based on the above I'd expect absolute paths for both to work:

```bash
$ docker build -t XXX --file `pwd`/Dockerfile `pwd`
```

But that gives the same "must be within build context" error.

Likewise based on a literal interpretation of the "PATH/Docker" in the help message, it would seem that `PATH=.` would be ok, but I get the same error with:

```bash
$ docker build -t XXX --file ./Dockerfile .
```

**Comment by duglin** [comment]
Do you use symlinks at all ?

-Doug

**Comment by rrnewton** [comment]
Nope, in fact in the same project checkout out on a mac and linux machines I'm getting different behavior.  Here's on my mac laptop:

```bash
$ docker --version
Docker version 1.8.1, build d12ea79
$ docker build -f Dockerfile .
unable to prepare context: The Dockerfile (/Users/.../Dockerfile) must be within the build context (.)
```

Here's on the linux server:

```bash
$ docker --version
Docker version 1.7.1, build 786b29d
$ docker build -f Dockerfile .
Sending build context to Docker daemon 22.44 MB
```

So maybe it's a 1.7-1.8 regression, or a problem with the Mac client.

**Comment by rrnewton** [comment]
Update: same problem on `Docker version 1.8.1, build d12ea79` on Linux.

**Comment by duglin** [comment]
@rrnewton I can't reproduce it, using 'master' and 'd12ea79'.  Are you sure there isn't anything special about your setup?

**Comment by p16** [comment]
Same problem here: 

```
$ docker build -t test -f DockerfileTest .
unable to prepare context: The Dockerfile (/path/to/project/DockerfileTest) must be within the build context (.)
```

docker version: `Docker version 1.8.1, build d12ea79`

**Comment by duglin** [comment]
geez this is weird - I just can't reproduce this.  Can someone make a tar or zip file available of a minimal dir structure?

**Comment by thaJeztah** [comment]
Wondering if all cases here are using a boot2docker setup and if that is related, also thinking things like OS X/Windows being case insensitive, whereas the Boot2Docker VM isn't. (I.e., I saw another (unrelated) example where `c:\some\path` didn't work, but `C:\some\path` did)

**Comment by duglin** [comment]
could be a boot2docker issue, but I don't think its a case insensitive thing because then it would generate a "file not found" type of error instead.

**Comment by thaJeztah** [comment]
doh, yes, should've switched on my brain before posting. Still interested to see if all here are using a boot2docker setup.

**Comment by rrnewton** [comment]
I was using boot2docker, but I aso got the same error on Linux with 1.8 but not 1.7.

**Comment by duglin** [comment]
@rrnewton any chance you could send me a tar/zip of a test dir that shows this?  And include a sh file showing the exact command you're using.

**Comment by ghudiczius** [comment]
I'm experiencing a similar problem, even when using full path for both the context and the Dockerfile

```
$ docker build -t project/machine:latest --rm=true -f $(pwd)/conf/base/machine/latest.Dockerfile $(pwd)/conf
unable to prepare context: The Dockerfile (d:\path\to\docker\files\conf\base\machine\latest.Dockerfile) must be within the build context (d:/path/to/docker/files/conf)
```

Note that while the path for the Dockerfile uses backslashes the path to the context uses slashes
This is on Windows 7 (64bit professional)

```
$ uname -a
MINGW32_NT-6.1 MACHINENAME 1.0.12(0.46/3/2) 2012-07-05 14:56 i686 unknown
```

```
$ docker version
Client:
 Version:      1.8.1
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   d12ea79
 Built:        Thu Aug 13 02:49:29 UTC 2015
 OS/Arch:      windows/amd64

Server:
 Version:      1.8.1
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   d12ea79
 Built:        Thu Aug 13 02:49:29 UTC 2015
 OS/Arch:      linux/amd64
```

```
$ docker -D info
Containers: 0
Images: 0
Storage Driver: aufs
 Root Dir: /mnt/sda1/var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 0
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.0.9-boot2docker
Operating System: Boot2Docker 1.8.1 (TCL 6.3); master : 7f12e95 - Thu Aug 13 03:24:56 UTC 2015
CPUs: 1
Total Memory: 1.956 GiB
Name: default
ID: UF45:XWKK:QGDY:G27T:IBXT:UZX2:7WYD:WFZK:YJXI:IKZB:57AI:IWZM
Debug mode (server): true
File Descriptors: 9
Goroutines: 16
System Time: 2015-08-18T09:43:49.241086399Z
EventsListeners: 0
Init SHA1:
Init Path: /usr/local/bin/docker
Docker Root Dir: /mnt/sda1/var/lib/docker
Labels:
 provider=virtualbox
```

**Comment by mingfang** [comment]
same problem using boot2docker.
build works in 1.7 but broken in 1.8.
qa anyone?

**Comment by wzheng2310** [comment]
Someone mentioned (or is guessing) if this is boot2docker issue. I don't know about the others but for me I am not using boot2docker.

**Comment by duglin** [comment]
just tried on a mac and I can't reproduce this.
does someone have time to send me a zip/tar of a test build that shows this?  I mean do you guys really have nothing but a Dockerfile in the current dir and nothing else? no links, nada?

**Comment by mingfang** [comment]
I created this repo to help reproduce the problem
https://github.com/mingfang/docker-test.git

**Comment by ghudiczius** [comment]
steps to reproduce (windows 7, powershell):

```
cd /
mkdir -p docker-14339
cd docker-14339
mkdir test
echo "FROM busybox" > test/test.Dockerfile

cmd /c mklink /d link test

docker build -f test/test.Dockerfile test
docker build -f "$(pwd)/test/test.Dockerfile" "$(pwd)/test"

docker build -f link/test.Dockerfile link
docker build -f "$(pwd)/link/test.Dockerfile" "$(pwd)/link"
```

the first two docker build command works, the third and fourth ones don't

**Comment by duglin** [comment]
@mingfang just tried it and it still works for me  :-(
@ghudiczius not sure what the 'mklink' command does but links with builds are a bit funky. I don't think we process softlinks which means a softlink to a Dockerfile outside of the build context won't work.

**Comment by ghudiczius** [comment]
`mklink /d` creates a directory sylink on windows (similar `ln -s`)
the above setup worked just fine with 1.7 and boot2docker

also on windows, when using mingw bash (git bash) the whole c:, d:, etc partitions are translated somehow (not sure though how) into /c/, /d/, etc, and i receive the same error as i mentioned in my initial post, which makes the mingw shell (git shell) unusuable

**Comment by mingfang** [comment]
I think I found the problem that I've been seeing.
It looks like the new boot2docker can only see files anywhere inside the home directory of the host;
It can not see files in /tmp, for example.

**Comment by duglin** [comment]
ping @tianon

**Comment by MJ-Lee** [comment]
The same with @mingfang 
I use a /tmp folder, get the same error.
Any follow up?

**Comment by thaJeztah** [comment]
> looks like the new boot2docker can only see files anywhere inside the home directory of the host;

Actually, I think that's always been the case; only the "Users" directory is (by default) shared with the Boot2Docker VM (using VirtualBox guest additions), any directory outside of "Users" must be manually added as a VBox share. See https://github.com/boot2docker/boot2docker/blob/master/README.md#virtualbox-guest-additions

**Comment by MJ-Lee** [comment]
@thaJeztah 
Actually, I am adding shared folder using command line:
VBoxManage sharedfolder add boot2docker-vm --name "Shared" --hostpath "folderInSystem"
and mount it inside boot2docker.
If there is a difference between using virtualbox guest additions and command line, I will let you know. 
Otherwise, I will post a new issue.

**Comment by michaeljs1990** [comment]
Same issue as everyone above.

- Docker toolbox / boot2docker
- Mac
- using tmp directory

**Comment by thaJeztah** [comment]
> Actually, I am adding shared folder using command line:
VBoxManage sharedfolder add boot2docker-vm --name "Shared" --hostpath "folderInSystem"
and mount it inside boot2docker.

ping @ehazlett @tianon I'm not too familiar with the boot2docker sharing; given the above comment, the user has manually added an extra share, but docker doesn't seem to pick that up; any ideas?

**Comment by tianon** [comment]
You have to mount additional shares manually (or via `bootlocal.sh`/`bootsync.sh`), or they will not be available inside the VM.

**Comment by MJ-Lee** [comment]
@tianon  @thaJeztah 
Hello, 
Yes, I already mounted it inside boot2docker vm(source code as following):
     sudo mkdir -p /shared
     sudo mount -t vboxsf -o rw,uid=1000,gid=1000 Shared /shared
and I can see the contents inside boot2docker vm. 
however when I try to docker build -f (and I create a /tmp folder where to create Dockerfile)
I got the issue: 
unable to prepare context: The Dockerfile (/xxx) must be within the build context (/xxx)

It's not a problem of shared system, I guess it's the docker1.8 cannot read the /tmp path for building with a file specified.

While docker 1.7 has no such problem

**Comment by 902Labs** [comment]
I am running an Ubuntu VM on OSX with VMWare Fusion.

My code directed is mounted and the following works just fine.

```
sudo docker build -f Dockerfile -t "$PKG" .
```

If I su into root and go into the same directory running the exact same command I get the context error listed above.

**Comment by jmanning2k** [comment]
docker 1.8.1 on OSX via docker-machine

When building using boot2docker or docker-machine in a symlinked folder (/tmp/repo symlinked to /User/username/src/repo):

```$ docker build -t XXX `pwd````

works just fine, as the symlink for pwd seems to get resolved locally, or at least before going into the VM.

However:

```$ docker build -t XXX --file Dockerfile `pwd````
```$ docker build -t XXX --file `pwd`/Dockerfile `pwd````

both fail, while using absolute paths (resolving symlinks) for both:

```$ docker build -t XXX --file `pwd -P`/Dockerfile `pwd -P````

is successful. This is definitely an issue of resolving symlinks for one path, while comparing with unresolved symlinks for another.

The error seen is:
```unable to prepare context: The Dockerfile (/tmp/repo/Dockerfile) must be within the build context (/tmp/repo)```

Please reopen - this is not the same as the linked issue.

**Comment by saurabhtandon13** [comment]
I was also getting the same error , actually it was a syntax error , can you please try to build the docker image using the below command

docker build -f <build file path> -t <docker image tag name>  < path of the output file once the build is complete>

**Comment by rncry** [comment]
Please reopen.. I'm having the same issue:
docker -v
Docker version 1.8.2, build 0a8c2e3

**Comment by rrnewton** [comment]
Just to confirm, @mingfang's `docker-test` reproducer does reproduce the problem for me on my mac:

```bash
$ docker --version
Docker version 1.8.1, build d12ea79
$ cd /tmp
$ git clone git@github.com:mingfang/docker-test.git
$ cd docker-test
$ docker build -f docker/Dockerfile -t test .
unable to prepare context: The Dockerfile (/tmp/docker-test/docker/Dockerfile) must be within the build context (.)
```

I notice that version is now out of date.  I'll try with the 1.8.2a release of Docker Toolbox.  The same recipe above now works fine for me on Linux, ubuntu 14.04 with `Docker version 1.8.2, build 0a8c2e3`.

**Comment by jmanning2k** [comment]
I've opened issue #16339 specifically targeting this issue, since this is already closed. Please add your comments there.

**Comment by berlincityboy** [comment]
I get the same error message. What does that mean for me now? I can't use Docker or what? What a cool technology :dancers:

**Comment by irbik** [comment]
Go inside the directory where you have got the Dockerfile:

1.- cd \folder\
2.- edit the file: Dockerfile

#INSTRUCTION FROM ubuntu:14.04
#INSTRUCTION MAINTAINER Kate Smith <iruiz@acens.com>
#INSTRUCTION RUN apt-get update && apt-get install -y ruby ruby-dev
#INSTRUCTION RUN gem install sinatra
FROM ubuntu:14.04
MAINTAINER Ivan Ruiz <iruiz@acens.com>
RUN apt-get update && apt-get install -y ruby ruby-dev
RUN gem install sinatra

3.-  Run -->  docker build -t ouruser/sinatra:v2 . ( Using the DOT "." for indicate the local path ).

I test with INSTRUCTION and does not work. Test with bouth.

I have the version 1.8.1

Regards

**Comment by thaJeztah** [comment]
@irbik not sure I understand; `INSTRUCTION` is not a Dockerfile command; https://docs.docker.com/engine/reference/builder/

**Comment by irbik** [comment]
Yes is true. The correct information inside the Dockfile is:

FROM ubuntu:14.04
MAINTAINER Ivan Ruiz iruiz@xxxxx.com
RUN apt-get update && apt-get install -y ruby ruby-dev
RUN gem install sinatra

**Comment by duglin** [comment]
@irbik I'm confused. This issue is about using "-f" but you're not using it.
Also, your Dockerfile works just fine for me.
What error are you seeing?

**Comment by irbik** [comment]
Sorry I can delete the wrong post. I got confused

**Comment by karlosmid** [comment]
On windows 7, and using docker-quickstart-terminal, followin cmd worked for me:

`docker build -t centos-node .`

**Comment by thaJeztah** [comment]
@karlosmid you're also not using the `-f` option so that's not related; please don't comment on closed issues, or at least make sure your issue is related to the issue.

**Comment by karlosmid** [comment]
@thaJeztah I posted workaround for not using -f option in the first place.

**Comment by singhniraj** [comment]
I was having same issue with dockers on window 7. The solution is to provide the the directory where docker will look for docker file.  

"docker build -f path/to/Dockerfile -t test ." This will look for docker file in the dir where the docker client exists. Instead you should provide 

"docker build -f path/to/Dockerfile -t test <<basedir of your dockerfile e.x d:/project/docker>>"

**Comment by insertinterestingnamehere** [comment]
I was having this issue as well when working with Windows Server containers. In my case, the dockerfile had inadvertently been created as `Dockerfile.txt` rather than just `Dockerfile`. The main issue wasn't symlinks. It was a bad filename and an unhelpful error message.

**Comment by hoshn** [comment]
@dddd

**Comment by Rahul91** [comment]
I was having the same issue, until the got the correct meaning of context. The relationship between Dockerfile and context, in programming terms can be explained with the concept of **scope** i.e. Dockerfile should be in the scope(just for our understanding) of context or as the error mentions context must contain Dockerfile.

You can have your Dockerfile anywhere in your home directory and you can use home as context
> /home/user/$ docker build -f somedir/somedir/somedir/somedir/Dockerfile . 

this will work because (.) i.e home directory contains Dockerfile

But if its reversed i.e.
> /home/user/$ docker build -f ~/Dockerfile  somedir/somedir/somedir/somedir/

It will throw the error and rightly so, as Dockerfile is not in context.

**Comment by elycruz** [comment]
Ha!  I had `DockerFile` instead of `Dockerfile` lol

**Comment by orangebook** [comment]
docker build [options] PATH | URL 使用Dockerfile构建镜像
 that PATH OR URL just like build Context，So just you alerady write --file "pwd"/path/dockerfile
but you don't choose buildcontext ,that also make it error

**Comment by jamesone** [comment]
Make sure your naming is correct! I had Dockefile :P This was also generated by IntelliJ

**Comment by tjtanvir666** [comment]
make sure the file is not Dockerfile.txt it should be Dockerfile wihtout any extension

**Comment by zored** [comment]
Why can't I specify Docker file from one location and context from another?

**Comment by thaJeztah** [comment]
@zored currently, the Dockerfile is sent to the daemon as part of the build-context, and the `--file` option only specifies where the daemon / builder can find the Dockerfile _inside_ the build-context (basically, the build-context is a `.tar` of the directory you specify as context).

This case came up recently in a discussion, and I (personally) don't see a real issue with the Dockerfile being pulled from "anywhere", and sent together with the build context (where it's put at the "root" of the build context during build), so e.g.;

```
docker build -t myapp -f /shared/dockerfiles/Dockerfile1 /projects/projecta/
```

In the above;

- `/projects/projecta/` is the build context to build from
- `/shared/dockerfiles/Dockerfile1` is the Dockerfile to use for building

The docker client would either send the Dockerfile separately, or include it in the build-context (.tar), and during build, the Dockerfile is put at the root of the build context.

There may be technical issues with implementing that, but I would not be against allowing that.

Actually: I just recalled where it came up as well; https://github.com/docker/docker/issues/27393

**Comment by asamuylikrps** [comment]
I would like to keep all my docker config files in one place rather than being scattered all over the project directory tree. They definitely should change that behaviour.

**Comment by AgentCoop** [comment]
I would like to keep all my docker config files in one place rather than being scattered all over the project directory tree. They definitely should change that behaviour.

**Comment by thaJeztah** [comment]
@AgentCoop perhaps you can open an issue / proposal / feature request for that?

**Comment by liuxuzxx** [comment]
My docker version is 1.6.2, My error info is :The Dockerfile (Dockerfile) must be within the build context (./jdk-tomcat) .   so how can i solve the problem?

**Comment by tmarshall** [comment]
I was running into the same issue, on OSX 10.12.4

```
$ docker --version
Docker version 17.06.1-ce, build 874a737
```

Problem happened because I am dynamically generating a Dockerfile, and placing it in a tmp dir. E.g. `/var/folders/r2/b4thjjrd0js3fyp091ydqy0w0000gn/T/tmp.N20VpqFA/base.Dockerfile`

Then:

```bash
DOCKERFILE_PATH="/var/folders/r2/b4thjjrd0js3fyp091ydqy0w0000gn/T/tmp.N20VpqFA/base.Dockerfile";
docker build -t "test" -f "$DOCKERFILE_PATH" .
```

^ this will fail, giving:

```
unable to prepare context: the Dockerfile (/var/folders/r2/b4thjjrd0js3fyp091ydqy0w0000gn/T/tmp.N20VpqFA/base.Dockerfile) must be within the build context
```

Though, changing my build command to this works:

```bash
DOCKERFILE_DIR="/var/folders/r2/b4thjjrd0js3fyp091ydqy0w0000gn/T/tmp.N20VpqFA/";
DOCKERFILE_PATH="/var/folders/r2/b4thjjrd0js3fyp091ydqy0w0000gn/T/tmp.N20VpqFA/base.Dockerfile";
docker build -t "test" -f "$DOCKERFILE_PATH" "$DOCKERFILE_DIR"
```

Which is fine for me, since the build context doesn't really matter in my case, but prob will cause issues for others.

**Comment by thaJeztah** [comment]
If you're dynamically generating the Dockerfile, there's no need to save it to a file first, just pipe it in through `stdin`

**Comment by tmarshall** [comment]
@thaJeztah yeah good point

**Comment by redbaron** [comment]
@thaJeztah ,

> @zored currently, the Dockerfile is sent to the daemon as part of the build-context, and the --file option only specifies where the daemon / builder can find the Dockerfile inside the build-context (basically, the build-context is a .tar of the directory you specify as context).

It contradicts documentation, which says (https://docs.docker.com/engine/reference/builder/#usage): 

> Traditionally, the Dockerfile is called Dockerfile and located in the root of the context. You use the -f flag with docker build to point to a Dockerfile anywhere in your file system.
```
$ docker build -f /path/to/a/Dockerfile .
```

**Comment by thaJeztah** [comment]
@redbaron good catch; looks like that's not properly documented; would you be interested in opening a pull request to fix that? the file used to create that part of the documentation can be found in the docker/cli repository:  https://github.com/docker/cli/blob/master/docs/reference/builder.md

**Comment by redbaron** [comment]
@thaJeztah , I'd think that if product behavior diverges from documentation, it is a bug in product requiring fixing :)

**Comment by thaJeztah** [comment]
See my earlier comment https://github.com/moby/moby/issues/14339#issuecomment-273540595 if you'd like to implement the code changes to match the incorrect documentation, that's fine with me as well :innocent:

**Comment by Nagendraking34** [comment]
"stderr": "unable to prepare context: unable to evaluate symlinks in Dockerfile path: lstat /opt/docker/Dockerfile: no such file or directory", "stderr_lines":

I am facing this issue

**Comment by thaJeztah** [comment]
@Nagendraking34 not sure if that's the same issue as is being discussed here (looks like there's a problem resolving the Dockerfile).

If you think there's a bug though, please open a new ticket instead, with details (as requested in the issue template when you open a new ticket) and exact steps to reproduce.

**Comment by Nagendraking34** [comment]
{"changed": true, "cmd": ["docker", "build", "-t", "simple-devops-image", "."], "delta": "0:00:00.040629", "end": "2021-12-09 15:12:56.379394", "msg": "non-zero return code", "rc": 1, "start": "2021-12-09 15:12:56.338765", "stderr": "unable to prepare context: unable to evaluate symlinks in Dockerfile path: lstat /home/ansadmin/Dockerfile: no such file or directory", "stderr_lines": ["unable to prepare context: unable to evaluate symlinks in Dockerfile path: lstat /home/ansadmin/Dockerfile: no such file or directory"], "stdout": "", "stdout_lines": []}


non zero return code they showing

**Comment by thaJeztah** [comment]
Please open a new ticket. Your comment only shows that "a build is failing"

```bash
docker build -t simple-devops-image .
```

but without more details (what's in the directory you're building from? does it have a Dockerfile in that directory? what platform are you running on? what docker version are you using? etc etc) it won't be possible for anyone to reproduce the issue to triage if there's a bug or not.
